--- a/common/main.cc	2023-01-08 18:19:49.873564419 -0500
+++ b/common/main.cc	2022-12-18 01:15:19.802959999 -0500
@@ -22,7 +22,8 @@
 #include <string>
 #include <cstdlib> // EXIT_SUCCESS, atoi()
 #include <getopt.h> // getopt_long
-
+#include <stdio.h> //Sleep
+#include <unistd.h> //Sleep
 #include "version.h"
 #include "graph.h"
 
@@ -245,13 +246,16 @@
   }
 
   MemoryStatus memory_status;
+  do {
   mem_status( memory_status );
   std::cout << mem_string( memory_status, mem_mode, use_colors, use_powerline_left, use_powerline_right )
     << cpu_string( cpu_mode, cpu_usage_delay, graph_lines, use_colors, use_powerline_left, use_powerline_right )
-    << load_string( use_colors, use_powerline_left, use_powerline_right, averages_count );
-
-  std::cout << std::endl;
+    << load_string( use_colors, use_powerline_left, use_powerline_right, averages_count )
+    << std::endl;
+//  sleep(2);
+  } while (1);
 
   return EXIT_SUCCESS;
 }
 
+
