Author: genr8eofl <genBTC@gmx.com>
Date: June 30, 2023
Subject: [PATCH] PORTDIR patch - change old to new portage tree directory
--- a/portconf	2014-02-03 07:04:59.000000000 -0500
+++ b/portconf	2023-06-30 19:59:01.779199114 -0400
@@ -62,7 +62,7 @@
 ################ all_use_all_profiles ######################################
 all_use_all_profiles(){
 	for target in `eselect --brief profile list`;do
-		profile_uses="$(profile "/usr/portage/profiles/${target}")"
+		profile_uses="$(profile "${PORTDIR}/profiles/${target}")"
 		USEs="$(sort_passed_uses "${profile_uses}")"
 		echo -e "\n${green}${target}${restore}: "${USEs} | fold -s -w `tput cols`
 		unset USEs
@@ -73,7 +73,7 @@
 profile_use(){
 	PS3=`echo -e "${contrast}Choose number:${restore} "`
 	select target in `eselect --brief profile list`;do
-		profile_uses="$(profile "/usr/portage/profiles/${target}")"
+		profile_uses="$(profile "${PORTDIR}/profiles/${target}")"
 		USEs="$(sort_passed_uses "${profile_uses}")"
 		echo -e "\n${green}${target}${restore}: "${USEs} | fold -s -w `tput cols`
 		unset USEs
@@ -88,7 +88,7 @@
 	else
 		target="$(readlink /etc/portage/make.profile | sed 's|.*/profiles/||')"
 	fi
-	profile_uses="$(profile "/usr/portage/profiles/${target}")"
+	profile_uses="$(profile "${PORTDIR}/profiles/${target}")"
 	USEs="$(sort_passed_uses "${PROFILE}")"
 	echo -e "${green}${target}${restore}: "${USEs} | fold -s -w `tput cols`
 	unset USEs
@@ -100,7 +100,7 @@
 . "${makefile}" 2>/dev/null
 MAKE_USES="${USE}"
 unset USE
-PORTDIR="${PORTDIR:-/usr/portage}"
+PORTDIR="${PORTDIR:-/var/db/repos/gentoo}"
 ############################################################################
 ################ timestamp #################################################
 timestamp(){
