--- /var/db/repos/gentoo/app-admin/mkosi/mkosi-17.1-r1.ebuild	2023-10-01 11:40:33.000000000 -0400
+++ /var/db/repos/myoverlay/app-admin/mkosi/mkosi-17.1-r2.ebuild	2023-11-23 13:53:24.740253224 -0500
@@ -4,7 +4,7 @@
 EAPI=8
 
 DISTUTILS_USE_PEP517=setuptools
-PYTHON_COMPAT=( python3_{9..11} )
+PYTHON_COMPAT=( python3_{9..12} )
 
 inherit distutils-r1 optfeature
 
@@ -20,20 +20,23 @@
 	sys-apps/bubblewrap
 	>=sys-apps/systemd-254
 	sys-firmware/edk2-ovmf"
-BDEPEND="virtual/pandoc"
+BDEPEND="
+    man? ( virtual/pandoc )
+"
+IUSE="man"
 
 distutils_enable_tests pytest
 
 src_compile() {
 	distutils-r1_src_compile
 
-	./tools/make-man-page.sh || die
+	use man && ./tools/make-man-page.sh
 }
 
 src_install() {
 	distutils-r1_src_install
 
-	doman mkosi/resources/mkosi.1
+	use man && doman mkosi/resources/mkosi.1
 }
 
 pkg_postinst() {
